### PR TITLE
Step data v2

### PR DESCRIPTION
**Pull request by Arpitjain111**

### CODE REVIEW

File: src/store/stepsV2/stepReducerV3.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import { setFunctionJsonThunkV3 } from './stepThunkV2'
+import { setBodyTypeAndValue } from './stepSelectorV2.ts'
+
+export const initialState: $StepsReduxType = {
+  isLoading: false
+}
+
+export const reducers: ValidateSliceCaseReducers<$StepsReduxType, SliceCaseReducers<$StepsReduxType>> = {
+  setKeyValueInstance(state, action) {
+    const tempInstance = cloneDeep(state[action.payload.currentScriptId][action.payload.currentStepId]?.draft)
```
Comment (rudrakachhawa) [inline src/store/stepsV2/stepReducerV3.ts:15]: prefer action destructuring here to use script id and step id

File: src/store/stepsV2/stepReducerV3.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    }
+  }
+}
+
+export function extraReducers(builder: ActionReducerMapBuilder<NoInfer<$StepsReduxType>>): void {
+  builder
+    // save User
+    .addCase(setFunctionJsonThunkV3.pending, (state) => {
+      state.isLoading = true
+    })
```
Comment (rudrakachhawa) [inline src/store/stepsV2/stepReducerV3.ts:34]: isLoading bhi paritcular step k liye true false krna he

File: src/store/stepsV2/stepReducerV3.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import { NoInfer } from 'react-redux'
+import { cloneDeep } from 'lodash'
+import { convertArrToString, makeUrl, urlWithParams } from '../../utils/utilities'
+import { $StepsReduxType } from '../../types/reduxCore.ts'
+import { setFunctionJsonThunkV3 } from './stepThunkV2'
+import { setBodyTypeAndValue } from './stepSelectorV2.ts'
+
+export const initialState: $StepsReduxType = {
+  isLoading: false
+}
```
Comment (rudrakachhawa) [inline src/store/stepsV2/stepReducerV3.ts:11]: api call abhi har bar step ke click par ho rahi he , uski wajah se draft wala code update ho jata he thats why mujhe saved code hi dikhta he , draft nahi dikhta he

File: src/components/masterSliderComponent/masterSliderComponent.tsx [hunk 4]
```diff
@@ -52,9 +53,29 @@ function MasterSliderComponent({ scriptId: currentScriptId, stepId: currentStepI
...
+          dispatch(setFunctionJsonThunkV3({ currentScriptId, currentStepId })).then(() => {
+            dispatch(
+              setKeyValueInstanceV2({
+                dataToUpdate: { title: flowJsonBlocks[currentSlugName]?.name || '' },
+                currentScriptId,
+                currentStepId
+              })
+            )
+          })
+        } else {
```
Comment (SiDDhartharG) [inline src/components/masterSliderComponent/masterSliderComponent.tsx:66]: .

Comment (SiDDhartharG) [reply src/components/masterSliderComponent/masterSliderComponent.tsx:66]: nagetion on function

File: src/components/responseAndConsoleTabs/responseAndConsoleTabs.tsx [hunk 5]
```diff
@@ -21,9 +21,8 @@ function TabPanel(props) {
   )
 }
 function ResponseAndConsoleTabs({ showPayloadTab, showConsoleTab, slugName: currentSlugName }) {
-  const { dryRunResponseForStep, lastInvocationData } = useCustomSelector((state: $ReduxCoreType) => ({
-    dryRunResponseForStep: state.stepsDataV2.dryRunResponseForStep,
-    lastInvocationData: state.invocationV2.invocationData
+  const { dryRunResponseForStep } = useCustomSelector((state: $ReduxCoreType) => ({
```
Comment (SiDDhartharG) [inline src/components/responseAndConsoleTabs/responseAndConsoleTabs.tsx:24]: .

Comment (SiDDhartharG) [reply src/components/responseAndConsoleTabs/responseAndConsoleTabs.tsx:24]: lastInvocationData

File: src/components/responseAndConsoleTabs/responseAndConsoleTabsV2.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  return (
+    // eslint-disable-next-line react/jsx-props-no-spreading
+    <div role='tabpanel' hidden={value !== index} id={`simple-tabpanel-${index}`} aria-labelledby={`simple-tab-${index}`} {...other}>
+      {value === index && <Box>{children}</Box>}
+    </div>
+  )
+}
+export default function ResponseAndConsoleTabs({ showPayloadTab, showConsoleTab }) {
+  const { currentSlugName, currentScriptId, currentStepId } = useSelector((state: $ReduxCoreType) => state.appInfo)
+  const dryRunResponseForStep = useSelector((state) => state.stepsDataV3?.[currentScriptId]?.[currentStepId]?.dryRunResponseForStep)
```
Comment (SiDDhartharG) [inline src/components/responseAndConsoleTabs/responseAndConsoleTabsV2.tsx:23]: useCustomSelector

File: src/components/saveButton/saveButton.tsx [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+import { flattenToJSON } from '../../store/plugin/currentSelectedPlugin/utility.ts'
+import { useCustomSelector } from '../../utils/deepCheckSelector'
+import { createFunctionThunk } from '../../store/stepsV2/stepThunkV2'
+
+export function SaveButton({ type, code, response, isDisabled }) {
+  const dispatch = useDispatch()
+  const orgId = useSelector((state) => state.appInfo.currentOrgId)
+  const projectId = useSelector((state) => state.appInfo.currentProjectId)
+  const currentScriptId = useSelector((state) => state.appInfo.currentScriptId)
+  const currentSlugName = useCustomSelector((state) => state.appInfo.currentSlugName)
```
Comment (SiDDhartharG) [inline src/components/saveButton/saveButton.tsx:21]: useCustomSelector

File: src/components/saveButton/saveButton.tsx [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+import { updateFunction } from '../../api'
+import { BlockTypes, CUSTOM_PLUGIN, Tabnames } from '../../enums'
+import { extractDomainWithoutWWW, replaceSpacesWithUnderscores } from '../../utils/utilities'
+import { setFlowJsonAndUsedVariables } from '../../store/flowJson/flowJsonSliceV2.ts'
+import { handleAddStepsV2 } from '../../store/invocationV2/invocationSliceV2.ts'
+import { flattenToJSON } from '../../store/plugin/currentSelectedPlugin/utility.ts'
+import { useCustomSelector } from '../../utils/deepCheckSelector'
+import { createFunctionThunk } from '../../store/stepsV2/stepThunkV2'
+
+export function SaveButton({ type, code, response, isDisabled }) {
```
Comment (SiDDhartharG) [inline src/components/saveButton/saveButton.tsx:16]: hoc

File: src/components/stepNameComponent/resetButtonForSlider.tsx [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import { useDispatch, useSelector } from 'react-redux'
+import RotateLeftIcon from '@mui/icons-material/RotateLeft'
+import { setFunctionJsonThunkV3 } from '../../store/stepsV2/stepThunkV2'
+import { setKeyValueInstance } from '../../store/stepsV2/stepSliceV3.ts'
+import { errorToast } from '../customToast'
+
+export default function ResetButtonForSlider() {
+  const dispatch = useDispatch()
+  const { currentScriptId, currentStepId, currentSlugName } = useSelector((state) => state.appInfo)
+  const flowJsonBlocks = useSelector((state) => state?.flowJsonV2?.flowJson?.blocks?.[currentSlugName])
```
Comment (SiDDhartharG) [inline src/components/stepNameComponent/resetButtonForSlider.tsx:11]: useCustomSelector

File: src/components/stepNameComponent/resetButtonForSlider.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
+import { Tooltip, IconButton, Box } from '@mui/material'
+import { useDispatch, useSelector } from 'react-redux'
+import RotateLeftIcon from '@mui/icons-material/RotateLeft'
+import { setFunctionJsonThunkV3 } from '../../store/stepsV2/stepThunkV2'
+import { setKeyValueInstance } from '../../store/stepsV2/stepSliceV3.ts'
+import { errorToast } from '../customToast'
+
+export default function ResetButtonForSlider() {
```
Comment (SiDDhartharG) [inline src/components/stepNameComponent/resetButtonForSlider.tsx:8]: HOC

File: src/components/stepNameComponent/stepNameComponentV2.tsx [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+import { Tabnames, sliderTypes } from '../../enums'
+import CustomFullScreen from '../customFullScreen.tsx'
+import { setKeyValueInstance } from '../../store/stepsV2/stepSliceV3.ts'
+import ResetButtonForSlider from './resetButtonForSlider.tsx'
+
+export function StepNameComponent({ showAskAiButton = true }) {
+  // author:@Arpit
+  const dispatch = useDispatch()
+  const currentSlugName = useSelector((state) => state.appInfo.currentSlugName)
+  const currentStepId = useSelector((state) => state.appInfo.currentStepId)
```
Comment (SiDDhartharG) [inline src/components/stepNameComponent/stepNameComponentV2.tsx:18]: useCustomSelector

File: src/components/stepNameComponent/stepNameComponentV2.tsx [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+import ApiIcon from '@mui/icons-material/Api'
+import SwitchSteps from '../common/switchStepsV2.tsx'
+import CloseSliderButton from '../common/closeSlidersButtonV2.tsx'
+import { openAiSlider, setShowFullScreen, toggleLogsSlider } from '../../store/appInfo/appInfoSlice.ts'
+import { Tabnames, sliderTypes } from '../../enums'
+import CustomFullScreen from '../customFullScreen.tsx'
+import { setKeyValueInstance } from '../../store/stepsV2/stepSliceV3.ts'
+import ResetButtonForSlider from './resetButtonForSlider.tsx'
+
+export function StepNameComponent({ showAskAiButton = true }) {
```
Comment (SiDDhartharG) [inline src/components/stepNameComponent/stepNameComponentV2.tsx:14]: HOC

File: src/store/combineReducer.js [hunk 13]
```diff
@@ -28,6 +29,7 @@ const rootReducer = combineReducers({
   appInfo: appInfoReducer,
   flowJsonV2: flowJsonReducer,
   stepsDataV2: stepsDataV2Reducer,
+  stepsDataV3: stepsDataV3Reducer,
```
Comment (SiDDhartharG) [inline src/store/combineReducer.js:32]: ADD in ReduxCore 
new data type

File: src/store/stepsV2/stepSliceV3.ts [hunk 14]
```diff
@@ -0,0 +1,16 @@
+import { SliceCaseReducers, createSlice } from '@reduxjs/toolkit'
+
+// // reducer imports
+import { $StepsReduxType } from '../../types/reduxCore.ts'
+import { initialState, reducers, extraReducers } from './stepReducerV3.ts'
+
+const stepsDataV3Slice = createSlice<$StepsReduxType, SliceCaseReducers<$StepsReduxType>, 'stepsDataV3'>({
+  name: 'stepsDataV3',
+  initialState,
```
Comment (SiDDhartharG) [inline src/store/stepsV2/stepSliceV3.ts:9]: data type if inti value

**Review by SiDDhartharG — CHANGES_REQUESTED** [review]

**Pull request closed without merging**